### ISSUE TITLE
[FEATURE] Skyline queries

### ISSUE THREAD

**Issue by MaxHalford** [koaning/scikit-lego#142]
Skyline queries are queries that find a subset of points that dominate the rest of the points but not the other members of the skyline. I wrote a [blog post](https://maxhalford.github.io/blog/skyline-queries-in-python/) on them. I also asked the `pandas` devs if this was in their scope but they said no. If this is something of interest to you I will gladly submit a pull request.

I'll describe another example in addition to my blog post. Imagine you're playing Mario Kart and want to find the best vehicle. Each vehicle has different characteristics, such as speed, weight, and maneuverability. Some vehicles will be slower but more maneuverable, etc. Some vehicles will also be objectively worse than at least one other vehicle because, say, they are slower, heavier, and less maneuverable. The skyline query will return all the "best" vehicles. In multi-objective this is also called the Pareto front.

If this is beyond your scope feel free to say so!

**Comment by koaning** [comment]
I'm actually kind of down with this idea. I am wondering though what part of the API it might fit in. It feels not unlike a monotonic regression except that we do not optimise for mean squared error but for furthers bound. 

There are many ways of interpolating this bound though; that's probably good to keep in mind. 

I would totally love to have this around, but I am wondering how to "score" this. 

@MBrouns do you have any thoughts on this?

**Comment by MaxHalford** [comment]
I'm pretty sure this should be a `pandas_util` function. It's not a transformer.

**Comment by MBrouns** [comment]
I'm a bit conflicted. On the one side I do like the functionality and could only find [pygmo](https://esa.github.io/pagmo2/docs/python/utils/py_mo_utils.html) with a reasonable implementation for this (also, I didn't know it but that library actually seems rather interesting). The other side though is that it is rather unrelated to the rest of the project and I'm not sure how much of a problem that is.

It could work as an `Estimator` that returns a binary vector indicating whether the point is on the pareto frontier

**Comment by koaning** [comment]
I like the idea of `Estimator`. 

I can imagine `Skyline().fit(X, y=None)` to learn where the current boundary is and `Skyline().fit(X, y=None).predict(X)` to tell you which points are `y=1` and which points aren't `y=0` on the efficient boundary. If a new dataset `X_new` is presented `Skyline().fit(X, y=None).predict(X_new)`it could do the same. 

One might even go as far as thinking about an interpretation of `.predict_proba()`. One might interpret a probability here as a distance measure between the current point and the effective boundary. Not sure if the word `_proba()` is going to be confusing here though. But again; I like the `Estimator` idea.

**Comment by MaxHalford** [comment]
Yeah I think `Estimator` is a good idea. Not so sure about `predict_proba` but maybe `decision_function` is a good idea.

**Comment by MBrouns** [comment]
`LogisticRegression` has a `decision_function` which makes sense indeed
